ui/raidboss/raidemulator.js

```diff
@@ -0,0 +1,832 @@
+        if (this.currentFight)
+          this.EndFight(log);
+        this.currentZone = logZoneChange;
+      }
+
+      if (this.currentFight) {
+        this.currentFight.logs.push(log);
+        if (this.IsWipe(log))
+          this.EndFight(log);
+      } else if (log.match(gLang.countdownEngageRegex())) {
+        // For consistency, only start fights on a countdown.  This
+        // makes it easy to know where to start all fights (vs
+        // reading timeline files or some such).
+        if (!this.currentZone)
+          console.error('Network log file specifies no zone?');
+        this.currentFight = {
+          zoneName: this.currentZone,
+          startDate: dateFromLogLine(log),
+          logs: [log],
+          durationMs: null,
+          key: this.fights.length,
+        };
+      }
+    }
+    this.StoreFights(this.fights);
+  }
+
+  EndFight(log) {
+    let endDate = dateFromLogLine(log);
+    let fight = this.currentFight;
+    fight.durationMs = endDate.getTime() - fight.startDate.getTime();
+    fight.endDate = endDate;
+
+    // Only add the fight if it hasn't been seen before.  The key
+    // generated internally by the collector is always unique so
+    // can't be used in this comparison.
+    this.currentFight = null;
+    for (let i = 0; i < this.fights.length; ++i) {
+      let f = this.fights[i];
+      if (f.logs[0] == fight.logs[0] && f.currentZone == fight.currentZone)
+        return;
+    }
+
+    // Trim log before pushing
+    // 00:0038 = echo | 00:0039 = countdown | 14, 15, 16 = attack related stuff
+    // 1A = gaining effects -> 1E losing effects | 1B = head Markers | 01 = zone change
+    // 03,04 = adding/removing combatant | 21 = wipe?
+    // Add more to regex only if needed
+    fight.logs = fight.logs.filter(function(line) {
+      return / 1[456AB]:| 0[134]:| 00:003[89]:| 21:.{8}:/.test(line);
+    }).sort();
+
+    this.fights.push(fight);
+    if (this.addFightCallback)
+      this.addFightCallback(fight);
+  }
+
+  IsWipe(log) {
+    // Actor control line list: https://gist.github.com/quisquous/250001cbce232a48e6a9ce772a56675a
+    return log.match(/ 21:........:40000010:/) || log.indexOf('00:0038:cactbot wipe') != -1;
+  }
+
+  ParseZoneChange(log) {
+    let m = log.match(/ 01:Changed Zone to (.*)\./);
+    if (!m)
+      return;
+    return m[1];
+  }
+}
+
+// Responsible for playing back a fight and emitting events as needed.
+class LogPlayer {
+  LogPlayer() {
+    this.Reset();
+  }
+
+  Reset() {
+    this.fight = null;
+    this.localStartTime = null;
+    this.logIdx = null;
+  }
+
+  SendLogEvent(logs) {
+    let evt = new CustomEvent('onLogEvent', { detail: { logs: logs } });
+    document.dispatchEvent(evt);
+  }
+
+  SendZoneEvent(zoneName) {
+    let evt = new CustomEvent('onZoneChangedEvent', { detail: { zoneName: zoneName } });
+    document.dispatchEvent(evt);
+  }
+
+  SendPlayerEvent(name, job, id) {
+    let jobDetail = {

```

Comment:
Do you need to send all of this?

Reply:
In an earlier version of my emulator I had the idea of having a party menu like the game itself showing HP etc. I will delete this on my next commit along with some other stuff.